test/module/irohad/ametsuchi/ametsuchi_test.cpp

```diff
@@ -809,3 +809,219 @@ TEST_F(AmetsuchiTest, TestRestoreWSV) {
+  shared_model::interface::Amount added_amount{"5.00"};
+  auto tx = shared_model::proto::TransactionBuilder()
+                .creatorAccountId("admin@test")
+                .createdTime(iroha::time::now())
+                .quorum(1)
+                .addAssetQuantity("coin#test", added_amount.toStringRepr())
+                .build()
+                .signAndAddSignature(key)
+                .finish();
+
+  auto block =
+      TestBlockBuilder()
+          .transactions(std::vector<shared_model::proto::Transaction>{tx})
+          .height(2)
+          .prevHash(genesis_block->hash())
+          .createdTime(iroha::time::now())
+          .build();
+
+  auto result = temp_wsv->apply(tx);
+  ASSERT_FALSE(framework::expected::err(result));
+  storage->prepareBlock(*temp_wsv);
+  temp_wsv.reset();
+
+  auto commited = storage->commitPrepared(block);
+
+  EXPECT_TRUE(commited);
+
+  shared_model::interface::Amount resultingAmount("10.00");
+
+  validateAccountAsset(
+      storage->getWsvQuery(), "admin@test", "coin#test", resultingAmount);
+}
+
+/**
+ * @given Storage with prepared state
+ * @when another block is applied
+ * @then state of the ledger is changed to that of the applied block
+ * and not of the prepared state
+ */
+TEST_F(PreparedBlockTest, PrepareBlockCommitDifferentBlock) {
+  // tx which we prepare
+  auto tx = shared_model::proto::TransactionBuilder()
+                .creatorAccountId("admin@test")
+                .createdTime(iroha::time::now())
+                .quorum(1)
+                .addAssetQuantity("coin#test", "5.00")
+                .build()
+                .signAndAddSignature(key)
+                .finish();
+
+  // tx which actually gets commited
+  auto other_tx = shared_model::proto::TransactionBuilder()
+                      .creatorAccountId("admin@test")
+                      .createdTime(iroha::time::now())
+                      .quorum(1)
+                      .addAssetQuantity("coin#test", "10.00")
+                      .build()
+                      .signAndAddSignature(key)
+                      .finish();
+
+  auto block =
+      TestBlockBuilder()
+          .transactions(std::vector<shared_model::proto::Transaction>{other_tx})
+          .height(2)
+          .prevHash(genesis_block->hash())
+          .createdTime(iroha::time::now())
+          .build();
+
+  auto result = temp_wsv->apply(tx);
+  ASSERT_TRUE(framework::expected::val(result));
+  storage->prepareBlock(*temp_wsv);
+  temp_wsv.reset();
+
+  apply(storage, block);
+
+  shared_model::interface::Amount resultingBalance{"15.00"};
+  validateAccountAsset(
+      storage->getWsvQuery(), "admin@test", "coin#test", resultingBalance);
+}
+
+/**
+ * @given Storage with prepared state
+ * @when another block is applied
+ * @then commitPrepared fails @and prepared state is not applied
+ */
+TEST_F(PreparedBlockTest, CommitPreparedFailsAfterCommit) {
+  // tx which we prepare
+  auto tx = shared_model::proto::TransactionBuilder()
```

Comment:
It looks like `tx`, `other_tx` can be moved to fixture and reused in test cases, and some shortcut for `block` can be created - a method which only takes a vector of transactions, for example.